README.md
## BULLETPROOF [![Build Status](https://travis-ci.org/samayo/bulletproof.svg?branch=master)](https://travis-ci.org/samayo/bulletproof.svg?branch=master)

[![Latest Stable Version](https://poser.pugx.org/samayo/bulletproof/v/stable.svg?format=flat-square)](https://packagist.org/packages/samayo/bulletproof) [![Total Downloads](https://poser.pugx.org/samayo/bulletproof/downloads?format=flat-square)](https://packagist.org/packages/samayo/bulletproof?format=flat-square) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/samayo/bulletproof/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/samayo/bulletproof/?branch=master)  [![Gitter chat](https://img.shields.io/badge/gitter-join--chat-blue.svg)](https://gitter.im/fastpress/fastpress) [![License](https://poser.pugx.org/samayo/bulletproof/license)](https://packagist.org/packages/fastpress/framework)

Bulletproof is a single-class library to upload images in PHP with security.

Install
-----

Using git
```bash
$ git clone https://github.com/samayo/bulletproof.git
```
Or composer
```bash
$ composer require samayo/bulletproof:4.0.*
```
Or [download it manually][bulletproof_archive] based on the archived version of release-cycles.

Usage
-----

Create an HTML form like this.
```html
<form method="POST" enctype="multipart/form-data">
  <input type="hidden" name="MAX_FILE_SIZE" value="1000000"/>
  <input type="file" name="pictures" accept="image/*"/>
  <input type="submit" value="upload"/>
</form>
```
And copy & paste the following code to upload the image
```php
require_once  "path/to/bulletproof.php";

$image = new Bulletproof\Image($_FILES);

if($image["pictures"]){
  $upload = $image->upload();

  if($upload){
    echo $upload->getFullPath(); // uploads/cat.gif
  }else{
    echo $image->getError();
  }
}
```
For more flexibility, check the options and examples below.


Configs
-----

#### Setting Properties
Before uploading, you can use these methods to restrict the image size, dimensions, mime types, location...
```php  
// Pass a custom name, or it will be auto-generated
$image->setName($name);

// define the min/max image upload size (size in bytes)
$image->setSize($min, $max);

// define allowed mime types to upload
$image->setMime(array('jpeg', 'gif'));

// set the max width/height limit of images to upload (limit in pixels)
$image->setDimension($width, $height);

// pass name (and optional chmod) to create folder for storage
$image->setLocation($folderName, $optionalPermission);
```

#### Getting Properties
Methods for getting image info before/after upload.
```php
// get the provided or auto-generated image name
$image->getName();

// get the image size (in bytes)
$image->getSize();

// get the image mime (extension)
$image->getMime();

// get the image width in pixels
$image->getWidth();

// get the image height in pixels
$image->getHeight();

// get image location (folder where images are uploaded)
$image->getLocation();

// get the full image path. ex 'images/logo.jpg'
$image->getFullPath();

// get the json format value of all the above information
$image->getJson();
```

#### Customized example
This will set image constrains and return output after upload
```php
$image = new Bulletproof\Image($_FILES);

$image->setName("samayo")
      ->setMime(["gif"])
      ->setLocation(__DIR__ . "/avatars");

if($image["pictures"]){
  if($image->upload()){
    echo $image->getName(); // samayo
    echo $image->getMime(); // gif
    echo $image->getLocation(); // avatars
    echo $image->getFullPath(); // avatars/samayo.gif
  }
}
```

#### Image Manipulation
To crop, resize or watermak images, use functions stored in [`src/utils`][utils]

#### Creating custom errors
Use php exceptions to define custom error responses
```php
if($image['pictures']){
  try {
    if($image->getMime() !== 'png'){
      throw new \Exception('Only PNG image types are allowed');
    }

    // check size, width, height...

    if(!$image->upload()){
      throw new \Exception($image->getError());
    } else {
      echo $image->getFullPath();
    }

  } catch (\Exception $e){
    echo "Error " . $e->getMessage();
  }
}
```

#### What makes this secure?  
* Uses **[`exif_imagetype()`][exif_imagetype_link]** to get the true image mime (`.extension`)
* Uses **[`getimagesize()`][getimagesize_link]** to check if image has a valid height / width in pixels.
* Sanitized images names, strict folder permissions and more...

### License: MIT
[utils]: https://github.com/samayo/bulletproof/tree/master/src/utils
[bulletproof_archive]: http://github.com/samayo/bulletproof/releases
[exif_imagetype_link]: http://php.net/manual/de/function.exif-imagetype.php
[getimagesize_link]: http://php.net/manual/en/function.getimagesize.php


CHANGELOG.md
# CHANGELOG

### 5.0.0
 - Removed built-in i18n to get rid of code bloat
 - lots of refractoring and changed method names

### 4.0.0
 - I'm not going to lie, most changes were done to get the perfect scrutinizer code quality
 - Also removed getting errors with `$image['error']` use `$image->getError()` instead
 - PHPunit requirement is ^6 - so tests for < php7.0 won't work (sorry)
 - Trying to upload images less or equal than 2px (width/height) is not longer prevented.
   Honestly, who am I to judge how many pixels an image should be?
 - Code is more refractored, it should be ready to eat now

### 3.2.0
- Fix some issues related to upload. (when file isn't uploaded, error wasn't shown)
- Fix PHPUnit namespace related error. (can't re-declare class .... error)
- Add .jpg to the list of default mime types
- re-write phpunit test class to use autoloader

### 3.0.0
 - revert back changes added by lordgiotto on v2.0.2 avoid complicating bulletproof
 - add improvements mentioned in issue #62
 - checking if exif_image exists is now done in the constructor instead of in upload() method
 - enable code coverage and quality inspection tool in scrutinizer
 - trying to upload .txt file would throw an error exposing directory path, now only error is displayed
 - code indetations and ps4 autoloader enabled

### 2.0.5
 - Add feature to get the mimetype before calling upload() method

### 2.0.4
 - fix #60 (JSON throws error due to a forwardslash in error messages)

### 2.0.3
 - git went haywire, 2.0.3 had to be born

### 2.0.2
 - enabled checking for 'exif_imagetype' function.
 - added configurable array messages thanks to github.com/lordgiotto

### 2.0.1
- Removed method `remove()`

### 2.0.0
- Removed watermark, resize, crop functionality
- renamed moaar and added getters and setters
- changed license to MIT
- added `getJson()` to get image info in json format
- enabled passing of the `$_FILES` array through the class constructor.
- wrote more unit tests
- class uses `\ArrayAccess` for intuitive file submit detection


test/uploadTest.php
<?php
declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use TestBootstrap\BulletproofTest;


class uploadTest extends TestCase {
    public $bulletproof;

    /**
     *  Initialize an array to mimic the properties $_FILES global
     */
    public function setUp() : void {
      $files = array(
          'ikea' => array(
              'name' => __DIR__ . "/monkey.jpg",
              'type' => 'image/jpg',
              'tmp_name' =>  __DIR__ . "/monkey.jpg",
              'error' => 0,
              'size' => 17438,
          )
      );

      $this->bulletproof = new BulletproofTest($files);
      $this->bulletproof['ikea'];

    }

    // check if the name provide in the html input is read by bulletproof
    public function testImageNameIsSame () {
      $this->bulletproof->setName('samayo');
      $this->assertEquals($this->bulletproof->getName(), 'samayo');
    }


    // test image accepts certain mime types
    public function testMimeTypes () {
      $this->bulletproof->setMime(['jpeg']);
      $upload = $this->bulletproof->upload();
      $this->assertEquals($upload->getMime(), 'jpeg');
    }

    // check dimensions params
    public function testDimensions () {
      // give it out of range (minimum dimentions than the class requires)
      $this->bulletproof->setDimension(1, 'b');
      $upload = $this->bulletproof->upload();
      $this->assertFalse($upload);
    }


    // check size of the image is correct
    public function testSize () {
      $this->bulletproof->upload();
      $this->assertEquals($this->bulletproof->getSize(), 17438); // 17438 is size of the monkey.jpg
    }

    // get mime of the image
    public function testMimeType() {
      $this->bulletproof->upload();
      $this->assertEquals($this->bulletproof->getMime(), 'jpeg');
    }

    // check width and height of the image
    public function testImageSizes() {
      $this->bulletproof->upload();
      $isWidth = $this->bulletproof->getWidth() === 384;
      $isHeight = $this->bulletproof->getHeight() === 345;
      $this->assertEquals($this->bulletproof->getMime(), 'jpeg');

      $this->assertEquals($isHeight, $isWidth);
    }

    // check if setting image storage is correct
    public function testLocation () {
      $this->bulletproof->setStorage('uploads');
      $this->assertEquals($this->bulletproof->getStorage(), 'uploads');
    }


    // check full path of image uploaded

   public function testFullpath(){
        $this->bulletproof->setStorage('uploads');
        $this->bulletproof->setName('2012');
        $this->bulletproof->setMime(['jpeg']);
        $upload = $this->bulletproof->upload();
        $this->assertSame($upload->getPath(), 'uploads/2012.jpeg');
    }


    // check json return value of image
     public function testJsonOutput(){
        $upload = $this->bulletproof->setName('we_belive_in_json')->upload();
        $this->assertSame($upload->getJson(),
            '{"name":"we_belive_in_json","mime":"jpeg","height":345,"width":384,"size":17438,"storage":"uploads","path":"uploads\/we_belive_in_json.jpeg"}');

    }

    // check invalid mimetype fails

    public function testMimeTypeFail () {
      $this->bulletproof->setMime(['gif']);
      $upload = $this->bulletproof->upload();
      $this->assertFalse($upload);
    }

    // check invalid mimetype fails with msg

    public function testMimeTypeFailWithMsg () {
      $this->bulletproof->setMime(['gif']);
      $upload = $this->bulletproof->upload();
      $this->assertEquals($this->bulletproof->getError(), 'Invalid File! Only (gif) image types are allowed');
    }

    // check invalid size with msg
    public function testImageSizeFailWithMsg () {
      $this->bulletproof->setSize(888, 9999);
      $upload = $this->bulletproof->upload();
       $this->assertEquals($this->bulletproof->getError(), 'Image size should be minimum 888 bytes (0 kb), upto maximum 9999 bytes (9 kb)');
    }


    // check invalid dimension fails with msg
public function testImageDimensionFailWithMsg () {
  $this->bulletproof->setDimension(42, 43);
  $this->bulletproof->upload();
    $this->assertEquals($this->bulletproof->getError(), 'Image height should be smaller than (43) pixels');
}



}


test/BulletproofTest.php
<?php

namespace TestBootstrap;

class BulletproofTest extends \Bulletproof\Image {

	/**
     * Return true at this point since we can't upload files
     * during test (or can we? I don't know!)
     */
    public function isSaved($tmp, $desination)
    {
        return true;
    }

    /**
     * Prevent class from making new folder
     */
    public function setLocation($dir = "bulletproof", $optionalPermision = 0666){
    	$this->location = $dir;
    	return $this;
    }
}


src/utils/func.image-crop.php
<?php
/**
 * bulletproof\utils\crop
 *
 * Image croping function for bulletproof library
 *
 * PHP support 5.3+
 *
 * @package     bulletproof
 * @version     3.2.0
 * @author      https://twitter.com/_samayo
 * @link        https://github.com/samayo/bulletproof
 * @license     MIT
 */
namespace Bulletproof\Utils;

function crop($image, $mimeType, $imgWidth, $imgHeight, $newWidth, $newHeight)
{
    switch ($mimeType) {
        case "jpg":
        case "jpeg":
            $imageCreate = imagecreatefromjpeg($image);
            break;

        case "png":
            $imageCreate = imagecreatefrompng($image);
            break;

        case "gif":
            $imageCreate = imagecreatefromgif($image);
            break;

        default:
            throw new \Exception(" Only gif, jpg, jpeg and png files can be cropped ");
            break;
    }

    // The image offsets/coordination to crop the image.
    $widthTrim = ceil(($imgWidth - $newWidth) / 2);
    $heightTrim = ceil(($imgHeight - $newHeight) / 2);

    // Can't crop to a bigger size, ex:
    // an image with 100X100 can not be cropped to 200X200. Image can only be cropped to smaller size.
    if ($widthTrim < 0 && $heightTrim < 0) {
        return;
    }

    $temp = imagecreatetruecolor($newWidth, $newHeight);
    imageAlphaBlending($temp, false);
    imageSaveAlpha($temp, true);
    imagecopyresampled(
        $temp,
        $imageCreate,
        0,
        0,
        $widthTrim,
        $heightTrim,
        $newWidth,
        $newHeight,
        $newWidth,
        $newHeight
    );


    if (!$temp) {
        throw new \Exception("Failed to crop image. Please pass the right parameters");
    }
        imagejpeg($temp, $image);
}


src/bulletproof.php
<?php
/**
 * BULLETPROOF.
 *
 * A single-class PHP library to upload images securely.
 *
 * PHP support 5.3+
 *
 * @version     5.0.0
 * @author      https://twitter.com/_samayo
 * @link        https://github.com/samayo/bulletproof
 * @license     MIT
 */
namespace Bulletproof;

class Image implements \ArrayAccess
{
    /**
     * @var string The new image name, to be provided or will be generated
     */
    protected $name;

    /**
     * @var int The image width in pixels
     */
    protected $width;

    /**
     * @var int The image height in pixels
     */
    protected $height;

    /**
     * @var string The image mime type (extension)
     */
    protected $mime;

    /**
     * @var string The full image path (dir + image + mime)
     */
    protected $path;

    /**
     * @var string The folder or image storage storage
     */
    protected $storage;

    /**
     * @var array The min and max image size allowed for upload (in bytes)
     */
    protected $size = array(100, 5000000);

    /**
     * @var array The max height and width image allowed
     */
    protected $dimensions = array(5000, 5000);

    /**
     * @var array The mime types allowed for upload
     */
    protected $mimeTypes = array('jpeg', 'png', 'gif', 'jpg');

    /**
     * @var array list of known image types
     */
    protected $acceptedMimes = array(
      1 => 'gif', 'jpeg', 'png', 'swf', 'psd',
      'bmp', 'tiff', 'tiff', 'jpc', 'jp2', 'jpx',
      'jb2', 'swc', 'iff', 'wbmp', 'xbm', 'ico'
    );

    /**
     * @var array error messages strings
     */
    protected $commonErrors = array(
        UPLOAD_ERR_OK => '',
        UPLOAD_ERR_INI_SIZE => 'Image is larger than the specified amount set by the server',
        UPLOAD_ERR_FORM_SIZE => 'Image is larger than the specified amount specified by browser',
        UPLOAD_ERR_PARTIAL => 'Image could not be fully uploaded. Please try again later',
        UPLOAD_ERR_NO_FILE => 'Image is not found',
        UPLOAD_ERR_NO_TMP_DIR => 'Can\'t write to disk, due to server configuration ( No tmp dir found )',
        UPLOAD_ERR_CANT_WRITE => 'Failed to write file to disk. Please check you file permissions',
        UPLOAD_ERR_EXTENSION => 'A PHP extension has halted this file upload process'
    );

    /**
     * @var array storage for the global array
     */
    private $_files = array();
    private $_file = array();

    /**
     * @var string storage for any errors
     */
    private $error = '';

    /**
     * @param array $_files represents the $_FILES array passed as dependency
     */
    public function __construct(array $_files = array())
    {

        /* check if php_exif is enabled */
        if (!function_exists('exif_imagetype')) {
          $this->error = 'Function \'exif_imagetype\' Not found. Please enable \'php_exif\' in your PHP.ini';
      }

      $this->_files = $_files;
    }

    /**
     * \ArrayAccess unused method
     *
     * @param mixed $offset
     * @param mixed $value
     */
    public function offsetSet($offset, $value){}

    /**
     * \ArrayAccess unused method
     *
     * @param mixed $offset
     */
    public function offsetExists($offset){}

    /**
     * \ArrayAccess unused method
     *
     * @param mixed $offset
     */
    public function offsetUnset($offset){}

    /**
     * \ArrayAccess - DOES NOT get array value from object
     *
     * This returns true/false based on the existence of an object at $offset.
     * It then does an ugly side effect of setting $this->_file to the object,
     * which affects the return value of other functions (getSize(), etc)
     *
     * @param mixed $offset
     *
     * @return bool
     */
    public function offsetGet($offset)
    {
        /* return false if $image['key'] isn't found */
        if (!isset($this->_files[$offset])) {
          $this->error = sprintf('No file input found with name: (%s)', $offset);

          return false;
        }

      $this->_file = $this->_files[$offset];

       /* check for common upload errors */
      if (isset($this->_file['error'])) {
        $this->error = $this->commonErrors[$this->_file['error']];
      }

      return $this->error ? false : true;
    }

    /**
     * Sets max image height and width limit.
     *
     * @param int $maxWidth max width value
     * @param int $maxHeight max height value
     *
     * @return $this
     */
    public function setDimension($maxWidth, $maxHeight)
    {
      $this->dimensions = array($maxWidth, $maxHeight);

      return $this;
    }

    /**
     * Returns the full path of the image ex 'storage/image.mime'.
     *
     * @return string
     */
    public function getPath()
    {
      return $this->path = $this->getStorage() . '/' . $this->getName() . '.' . $this->getMime();
    }

    /**
     * Returns the image size in bytes.
     *
     * @return int
     */
    public function getSize()
    {
      return (int) $this->_file['size'];
    }

    /**
     * Define a min and max image size for uploading.
     *
     * @param int $min minimum value in bytes
     * @param int $max maximum value in bytes
     *
     * @return $this
     */
    public function setSize($min, $max)
    {
      $this->size = array($min, $max);
      return $this;
    }

    /**
     * Returns a JSON format of the image width, height, name, mime ...
     *
     * @return string
     */
    public function getJson()
    {
      return json_encode(
        array(
          'name' => $this->name,
          'mime' => $this->mime,
          'height' => $this->height,
          'width' => $this->width,
          'size' => $this->_file['size'],
          'storage' => $this->storage,
          'path' => $this->path,
        )
      );
    }

    /**
     * Returns the image mime type.
     *
     * @return null|string
     */
    public function getMime()
    {
      if (!$this->mime) {
        $this->mime = $this->getImageMime($this->_file['tmp_name']);
      }

      return $this->mime;
    }

    /**
     * Define a mime type for uploading.
     *
     * @param array $fileTypes
     *
     * @return $this
     */
    public function setMime(array $fileTypes)
    {
      $this->mimeTypes = $fileTypes;
      return $this;
    }

    /**
     * Gets the real image mime type.
     *
     * @param string $tmp_name The upload tmp directory
     *
     * @return null|string
     */
    protected function getImageMime($tmp_name)
    {
      $this->mime = @$this->acceptedMimes[exif_imagetype($tmp_name)];
      if (!$this->mime) {
        return null;
      }

      return $this->mime;
    }

    /**
     * Returns error string
     *
     * @return string
     */
    public function getError()
    {
      return $this->error;
    }

    /**
     * Returns the image name.
     *
     * @return string
     */
    public function getName()
    {
      return $this->name;
    }

    /**
     * Provide image name if not provided.
     *
     * @param string $isNameProvided
     *
     * @return $this
     */
    public function setName($isNameProvided = null)
    {

      if ($isNameProvided) {
        $this->name = filter_var($isNameProvided, FILTER_SANITIZE_STRING);
      }else{
        $this->name = uniqid('', true) . '_' . str_shuffle(implode(range('e', 'q')));
      }

      return $this;
    }

    /**
     * Returns the image width.
     *
     * @return int
     */
    public function getWidth()
    {
      if ($this->width != null) {
        return $this->width;
      }

      list($width) = getimagesize($this->_file['tmp_name']);

      return $width;
    }

    /**
     * Returns the image height in pixels.
     *
     * @return int
     */
    public function getHeight()
    {
      if ($this->height != null) {
        return $this->height;
      }

      list(, $height) = getimagesize($this->_file['tmp_name']);

      return $height;
    }

    /**
     * Returns the storage / folder name.
     *
     * @return string
     */
    public function getStorage()
    {
      if (!$this->storage) {
        $this->setStorage();
      }

      return $this->storage;
    }

    /**
     * Validate directory/permission before creating a folder.
     *
     * @param string $dir the folder name to check
     *
     * @return bool
     */
    private function isDirectoryValid($dir)
    {
      return !file_exists($dir) && !is_dir($dir) || is_writable($dir);
    }

    /**
     * Sets storage for upload storage.
     *
     * @param string $dir the folder name to create
     * @param int $permission chmod permission
     *
     * @return $this
     */
    public function setStorage($dir = 'uploads', $permission = 0666)
    {
      $this->storage = $this->createStorage($dir, $permission);
      return $this;
    }

    /**
     * Creates a storage for upload storage.
     *
     * @param string $dir the folder name to create
     * @param int $permission chmod permission
     *
     * @return string?
     */
    public function createStorage($dir = 'uploads', $permission = 0666)
    {
      $isDirectoryValid = $this->isDirectoryValid($dir);

      if (!$isDirectoryValid) {
        $this->error = 'Can not create a directory  \''.$dir.'\', please check write permission';
        return false;
      }

      $create = !is_dir($dir) ? @mkdir('' . $dir, (int) $permission, true) : true;

      if (!$create) {
        $this->error = 'Error! directory \'' . $dir . '\' could not be created';
        return false;
      }

      return $dir;
    }


    public function validateMime() {
      $this->getImageMime($this->_file['tmp_name']);

      if(!in_array($this->mime, $this->mimeTypes)) {
        $this->error = sprintf('Invalid File! Only (%s) image types are allowed', implode(', ', $this->mimeTypes));
        return false;
      }

      return true;
    }

    public function validateSize() {
      list($minSize, $maxSize) = $this->size;
      if ($this->_file['size'] < $minSize || $this->_file['size'] > $maxSize) {
        $min = $minSize.' bytes ('.intval($minSize / 1000).' kb)';
        $max = $maxSize.' bytes ('.intval($maxSize / 1000).' kb)';
        $this->error = 'Image size should be minimum '.$min.', upto maximum '.$max;
        return false;
      }

      return true;
    }

    public function validateDimension() {
      /* check image dimension */
      list($maxWidth, $maxHeight) = $this->dimensions;
      $this->width = $this->getWidth();
      $this->height = $this->getHeight();

      if($this->height > $maxHeight) {
        $this->error = sprintf('Image height should be smaller than (%s) pixels', $maxHeight);

        return false;
      }

      if($this->width > $maxWidth) {
        $this->error = sprintf('Image width should be smaller than (%s) pixels', $maxHeight);

        return false;
      }

      return true;
    }

    public function isValid() {

      return $this->validateMime() && $this->validateSize() && $this->validateDimension();
    }


    /**
     * Validate and save (upload) file
     *
     * @return false|Image
     */
    public function upload()
    {

      if ($this->error !== '') {
        return false;
      }

      $this->getName();


      $isSuccess = $this->isValid() && $this->isSaved($this->_file['tmp_name'], $this->getPath());

      return $isSuccess ? $this : false;
    }

    /**
     * Final upload method to be called, isolated for testing purposes.
     *
     * @param string $tmp_name the temporary storage of the image file
     * @param string $destination upload destination
     *
     * @return bool
     */
    protected function isSaved($tmp_name, $destination)
    {
      return move_uploaded_file($tmp_name, $destination);
    }
}
